django/core/management/commands/runserver.py

```diff
@@ -135,23 +122,39 @@ def inner_run(self, *args, **options):
             "settings": settings.SETTINGS_MODULE,
             "protocol": self.protocol,
             "addr": '[%s]' % self.addr if self._raw_ipv6 else self.addr,
-            "port": self.port,
+            "port": server_port,
             "quit_command": quit_command,
         })
 
+    def inner_run(self, *args, **options):
+        # If an exception was silenced in ManagementUtility.execute in order
+        # to be raised in the child process, raise it now.
+        autoreload.raise_last_exception()
+
+        threading = options['use_threading']
+        # 'shutdown_message' is a stealth option.
+        shutdown_message = options.get('shutdown_message', '')
+
+        if not options['skip_checks']:
+            self.stdout.write('Performing system checks...\n\n')
+            self.check(display_num_errors=True)
+        # Need to check migrations here, so can't use the
+        # requires_migrations_check attribute.
+        self.check_migrations()
         try:
             handler = self.get_handler(*args, **options)
             run(self.addr, int(self.port), handler,
-                ipv6=self.use_ipv6, threading=threading, server_cls=self.server_cls)
+                ipv6=self.use_ipv6, threading=threading,
+                server_cls=self.server_cls, on_bind=self._on_bind)
         except OSError as e:
             # Use helpful error messages instead of ugly tracebacks.
             ERRORS = {
-                errno.EACCES: "You don't have permission to access that port.",
-                errno.EADDRINUSE: "That port is already in use.",
-                errno.EADDRNOTAVAIL: "That IP address can't be assigned to.",
+                errno.EACCES: "You don't have permission to access port %(port)s.",
+                errno.EADDRINUSE: "Port %(port)s is already in use.",
+                errno.EADDRNOTAVAIL: "IP address %(addr)r can't be assigned to.",
```

Comment:
We could just use f-strings here:

```suggestion
                errno.EACCES: f"You don't have permission to access port {self.port}.",
                errno.EADDRINUSE: f"Port {self.port} is already in use.",
                errno.EADDRNOTAVAIL: f"IP address {self.addr!r} can't be assigned to.",
```

(I know this means generating three strings when we'll only show one, but it isn't really going to add any significant overhead.)